Move ARC data buffers out of vmalloc

tuxoko opened the pull request:
***Warning***
This patchset requires further testing and review.
DO NOT USE IN PRODUCTION.

---
Introduce ABD: linear/scatter dual typed buffer for ARC

zfsolinux currently uses vmalloc backed slab for ARC buffers. There are some
major problems with this approach. One is that 32-bit system have only a
handful of vmalloc space. Another is that the fragmentation in slab will easily
trigger OOM in busy system.

With ABD, we use scatterlist to allocate data buffers. In this approach we can
allocate in HIGHMEM, which alleviates vmalloc space pressure on 32-bit. Also,
we don't have to rely on slab, so there's no fragmentation issue.

But for metadata buffers, we still uses linear buffer from slab. The reason for
this is that there are a lot of *_phys pointers directly point to metadata
buffers. So it's kind of impractical to change all those code. And metadata
should only be a small percentage so we probably don't have to worry about
the 32-bit and fragmentation issue, right?

---
This patchset is divided into the following part:
1. Add compatibility layer for {kmap,kunmap}\_atomic
2. Introduce ABD, but with scatter type disabled
3. Modify checksum implementation for ABD
4. Modify every file which touches abd_t field
5. Enable scatter type ABD

---
Edit:
I forgot to mention...
There are some parts that hasn't been able to handle scatter buffer,
compression and raidz gaussian elimination to name a few.
Enable them will cost you extra copy than without this patch set.
Another thing is the xuio thing, which I totally skip.

Todo:
1. Clean up the implementation and documentation
2. Compression and raidz gaussian elimination takes scatter buffer
3. Probably change vdev_cache and vdev_queue to scatter buffer
4. vdev_file should be able to submit a iovec.

greg-hydrogen commented:
@tuxoko any chance you can update this patch stack?

tuxoko commented:
@behlendorf 
Thanks for the notice.
I'll look into the kmem stuff and see what's coming.
Hopefully, I'll start work on this series as soon as you finish.

prakashsurya commented:
@tuxoko @behlendorf Can one or both of you quickly comment on the state of this patch? Is it close to being "production ready"? are there any known issue that remain to be fixed? how much testing has it undergone so far (I've seen people say they run with it applied, but I'm unsure of the extent of this testing)? etc..

I ask because I might be tasked with pulling this work into illumos for our next Delphix release. So, I'm going to try and evaluate how much work that'd actually take. Assuming I get the green light to work on getting this landed in illumos, what would be the best strategy? Should I start by porting this to illumos with the idea that ZoL will pull it down once it lands in illumos (and we can collaborate through illumos review comments)? Or maybe it'd be best to just merge this work to ZoL first and I can pull it into illumos from there?

Obviously my deliverable would be to get a patch landed on illumos, but I think it's in everybody's best interested to develop a solution that is cross-platform. Let me know what you guys think on how to proceed.

tuxoko commented:
I updated this patch series.
It turned out not many conflicts needed to be dealt with.

Also, the panic above hit by @DeHackEd emerges when you directly rebase the old ABD revision onto the master. The problematic part is exactly where the compiler warning occurs as @sempervictus pointed out. So it would be a good idea not to ignore them.
Anyway this issue is dealt with in ```restore_write()```. So it shouldn't happen again.

tuxoko commented:
@prakashsurya 
I personally think this patch series is quite stable. However, there are some corners that should be smoothed out before being merged. I'll work on them once the kmem-rework mentioned by @behlendorf is finished.

prakashsurya commented:
@tuxoko Excellent, I've been skimming over the patch the past hour or so, and they look quite nice! How they're all split up into discrete chunks makes it much easier to make sense of it all. Can you elaborate a bit on the sharp corners?

If I'm understanding this right, the main entry point is `abd_alloc()` which is replacing the existing large `zio_buf_alloc()` calls? And then there's a lot of changes surrounding the correct usage and management of these scatter gather lists?

Looking this over, I'm encouraged that it won't be too difficult of an endeavor to get these working on illumos. Off the top of your head, is there any Linux specific code that I need to pay attention to?

sempervictus commented:
@prakashsurya: in terms of testing, this patch set is usually included in our production builds for our storage servers as well as personal desktop/laptop builds for engineers. In both cases the performance benefits are significant, and all builds are run through ztest, fio, and tiotest before getting thrown into the file/block-server deployment stack. We run XenServer and VMWare clusters atop the iSCSI exports in production environments, and have Arista 10G-backed SSD pools serving iSCSI blocks (512) consistently @ ~600MB/s from 3disk Z1 pools... Long-term stable builds with this patch set have outperformed upstream consistently in iSCSI, NFS, and desktop workloads.

Great to hear Illumos may be going in this direction - i dont know if this will help as much there, where the native memory management code is well attuned to ZFS, but it will be nice to have the FS implementation decoupled from the design decisions made in another kernel (from a Linux user's perspective). Given that the patch stack was written for ZoL, i would think it would be better to have it achieve its goals in this tree, then port and merge the work upstream such that all consumers can make use of it down the line. Implementing the "master version" in Illumos may cause changes to creep in which have adverse affects on the initially desired target platform.

@tuxoko: thanks for yet another update, i'll get this pushed out to the test mules tonight. I've been using the kmem rework code recently, and still seeing some merge conflicts from a clean master + 2918 + this (vdev_queue.c, sa.c, arc.c in module/zfs). Will take a look at those before i build for testing as well. Since master will soon have 2918, is there any chance i could ask you could perform the merge and create a branch for it in your repo? Every time i fix a merge conflict in ZFS (especially @ lower layers) i get mental imagery of data in a meat grinder after the next module reload.

prakashsurya commented:
@sempervictus Thanks for the info! That's encouraging to hear it's being tested with production workloads and is holding up. And I agree, I'm all for this work landing in ZoL as soon as possible.

While the problem is slightly different on illumos, the solution (using scatter/gather) is the same; which is why I'm looking at pulling this over. Since this work is eventually going to land in ZoL no matter what (I presume from previous conversations with @behlendorf), it also has the benefit of reducing the differences between the two platforms, which I think we'd want to do regardless of whether it had any other tangible benefit on illumos.

With that said, it would be good to determine a good way to collaborate and share any changes that need to be made during the review process; I likely won't be able to wait for this to solidify and land in ZoL before I start pulling things into illumos.

sempervictus commented:
@prakashsurya given that Illumos is OpenZFS' source of truth for most matters, i would guess that s/g changes made upstream as a result of this work would subsequently trickle back down to ZoL once Illumos adapts and commits anyway. That would require changes to the same code in master after initial commit, but if functionality is preserved, my vote is "worth it to have this functionality in ZoL sooner."

@tuxoko: you've been gracious enough to submit and maintain this stack for almost a year. Would you prefer to keep updating the PR to try and get this work into Illumos/upstream first with greater consistency/review/testing/etc, or push the commit to ZoL master sooner and address those issues post-commit?

tuxoko commented:
@prakashsurya 
The most ugliest thing in this patch series is probably the last bit hack for determining scatter or linear type buffer. It should be hide behind the ABD struct and APIs.
Although it's not that hard to fix (and the main reason it hasn't been done is due to my laziness), what I really like to do  is to use ```scatterlist``` for both scatter buffer and linear buffer, so we only need single version of function for each API. However, earlier version of Linux doesn't provide "skip" for the ```sg_miter``` iterator, which would make everyone's life much easier when looking at the code. So, I'm kind of hesitate which approach I should use.

Other problems like making things work with scatter buffer, so we don't need extra copy. The hardest part for this would be the compression algorithms. But these could be dealt afterward if it's acceptable with everyone.

kernelOfTruth commented:
FYI:

Kmem rework has been merged:

https://github.com/zfsonlinux/zfs/pull/2918#event-220411321
https://github.com/zfsonlinux/spl/pull/414#event-220411214

https://github.com/zfsonlinux/zfs/commit/6e9710f7c33a7440904b39fb4e6b68b3325cfeca
https://github.com/zfsonlinux/spl/commit/9099312977e994f8638eb1d93fd92a5cd66b8ea6


@tuxoko feel free to take all the time you need to work on it,
from what I can recall - I haven't had any stability or other related issues with this patchset so far

when it's been ported to work with the kmem-rework changes I'll probably start to give ZFS some wider stability/performance testing (e.g. for the portage-tree [Gentoo/Linux], /var/tmp/portage [temporary compilation folder], etc.)

Thanks ! :+1:

tuxoko commented:
Hi all,
I just push out a new revision of the patch series.
This is a rather major update, so there could be new bug introduced, though I didn't find any in my testing. Anyway, test it before using it.

Notable changes are:
1. Both scatter and linear type ABD are now under the ABD struct. No more bit hack (finally!)
2. In stead of using the sg_miter provided by kernel, I introduced my own abd_miter. The reason for this is that a) it can be use on both type of ABD, b) sg_miter cannot be used to simultaneously iterate on two scatterlist in older kernel, and c) we sometimes need to fall back to non-atomic kmap.
3. Some API changes.

Now with this version being pushed out. I believe this patch series is ready for serious review.

greg-hydrogen commented:
Thanks @tuxoko for updating this patch stack

@all - I have a new server that I can put this patch to the test... is there any testing methodology that would help uncover any potential issues?

tuxoko commented:
The version I pushed yesterday would fail to build on older kernels.
This one should fix the problem.

sempervictus commented:
Built in the patch stack with no issues:
```
  * origin/pr/3078
  * origin/pr/3074
  * origin/pr/2970
  * origin/pr/2909
  * origin/pr/2890
  * origin/pr/2820
  * origin/pr/2784
  * origin/pr/2012
  * origin/pr/1099
  * origin/pr/2685
  * origin/pr/3063
  * origin/pr/3076
  * origin/pr/2129
  * master @ bf5efb5c66ac30442bded92c3299db36fe21d92f
```
Quick set of benchmarks makes me think we've got a bottleneck in the compression routines somewhere (testing on LZ4 and without).

Without LZ4 on a 1T SSD (no atime no xattr no acl):
```
Tiotest results for 8 concurrent io threads:
,----------------------------------------------------------------------.
| Item                  | Time     | Rate         | Usr CPU  | Sys CPU |
+```-----------------------+----------+--------------+----------+---------+
| Write        1600 MBs |   27.8 s |  57.615 MB/s |  98.7 %  | 101.0 % |
| Random Write   31 MBs |    0.5 s |  56.828 MB/s | 318.4 %  | 169.8 % |
| Read         1600 MBs |    0.5 s | 2926.972 MB/s | 3103.9 %  | 1874.4 % |
| Random Read    31 MBs |    0.0 s | 2710.791 MB/s | 2966.7 %  | 1873.7 % |
`----------------------------------------------------------------------'
Tiotest latency results:
,-------------------------------------------------------------------------.
| Item         | Average latency | Maximum latency | % >2 sec | % >10 sec |
+--------------+-----------------+-----------------+----------+-----------+
| Write        |        0.019 ms |        2.928 ms |  0.00000 |   0.00000 |
| Random Write |        0.074 ms |        9.375 ms |  0.00000 |   0.00000 |
| Read         |        0.009 ms |        0.228 ms |  0.00000 |   0.00000 |
| Random Read  |        0.009 ms |        0.091 ms |  0.00000 |   0.00000 |
|--------------+-----------------+-----------------+----------+-----------|
| Total        |        0.015 ms |        9.375 ms |  0.00000 |   0.00000 |
`--------------+-----------------+-----------------+----------+-----------'

```
Note the read speeds there approaching 3GB/s, this is what i used to see before the kmem changes started coming in and i had to pull ABD out of the patch stack, used to look like this with or without LZ4.

Here's the same test with LZ4 enabled:
```
Tiotest results for 8 concurrent io threads:
,----------------------------------------------------------------------.
| Item                  | Time     | Rate         | Usr CPU  | Sys CPU |
+-----------------------+----------+--------------+----------+---------+
| Write        1600 MBs |    2.0 s | 809.895 MB/s | 1234.3 %  | 1448.1 % |
| Random Write   31 MBs |    0.4 s |  88.894 MB/s | 382.0 %  | 346.2 % |
| Read         1600 MBs |    1.1 s | 1477.113 MB/s | 763.7 %  | 3974.8 % |
| Random Read    31 MBs |    0.0 s | 1441.022 MB/s | 848.5 %  | 3827.4 % |
`----------------------------------------------------------------------'
Tiotest latency results:
,-------------------------------------------------------------------------.
| Item         | Average latency | Maximum latency | % >2 sec | % >10 sec |
+--------------+-----------------+-----------------+----------+-----------+
| Write        |        0.019 ms |        2.269 ms |  0.00000 |   0.00000 |
| Random Write |        0.045 ms |        3.415 ms |  0.00000 |   0.00000 |
| Read         |        0.020 ms |        0.281 ms |  0.00000 |   0.00000 |
| Random Read  |        0.020 ms |        0.059 ms |  0.00000 |   0.00000 |
|--------------+-----------------+-----------------+----------+-----------|
| Total        |        0.020 ms |        3.415 ms |  0.00000 |   0.00000 |
`--------------+-----------------+-----------------+----------+-----------'
```
That 1400MB/s number is too close (1479/1463 on last record) to what i was seeing a few days ago without ABD applied, and smells synthetic as hell to me now that i'm looking at both benchmarks. 

That said, i've got an iSCSI node opening up at the datacenter which i can dedicate to the lab account in our cloud in order to keep production VMs off the sucker. SCST tends to bring out the worst in ZFS far as i can tell, so we'll see what awful stack traces it'll come up with.

behlendorf commented:
@tuxoko thank you for continuing to push this work forward it looks _very_ nice!  You've done a great job breaking up the patch stack and documenting things to make it easier to review.

I definitely want to get this in ZoL in the relatively near future.  What I'd like to see happen is for this work to be ready to merge immediately after the next ZoL tag.  Likely right after the large block patches are merged to minimize any upstream differences with Illumos.  Since it's a big change it's inevitable that it will be slightly destabilizing no matter how much it's tested so the sooner it starts seeing real workloads the better.

@prakashsurya it would be awesome if you could champion this on the Illumos side.  If there are changes which need to be made to make it more Illumos friendly it would be good to understand what they are now so they can be addressed right away.

edillmann commented:
FYI @tuxoko it seems that this patchset breaks spacemap accounting
A simple zpool create followed by a zdb -mc shows spacemap problems with this PR applied (abd_next)

tuxoko commented:
@edillmann 
>A simple zpool create followed by a zdb -mc shows spacemap problems with this PR applied

Does this always occur to you. I've never encountered it during my testing.

edillmann commented:
@tuxoko at least with master+abd_next & dual vdevs it's occuring always

It did the same test with master which doesn't show this problem

edillmann commented:
@tuxoko I dig a little bit more into this and tried to compile master on an host running abd_test

zdb -mc gave me the following result :

Traversing all blocks to verify metadata checksums and verify nothing leaked ...

loading space map for vdev 0 of 1, metaslab 173 of 174 ...
33.0G completed (1074MB/s) estimated time remaining: 0hr 00min 00sec        
        No leaks (block sum matches space maps exactly)

        bp count:          791056
        bp logical:    55452135424      avg:  70098
        bp physical:   31266612736      avg:  39525     compression:   1.77
        bp allocated:  35669655552      avg:  45091     compression:   1.55
        bp deduped:             0    ref>1:      0   deduplication:   1.00
        SPA allocated: 35669655552     used: 38.18%

        additional, non-pointer bps of type 0:      54979

space map refcount mismatch: expected 175 != actual 174

So perhaps is the bug only on the zdb tool side ...

prakashsurya commented:
@behlendorf I'll see what I can do. The project that involved incorporating this patch-set got tabled for our upcoming release, so it's not a sure bet that I'll be able to give this sufficient time in the short term. With that said, Matt and I are both interested in giving this a closer look, so I'll try and squeeze this in the best I can.

FWIW, using a platform agnostic scatter gather implementation sounds like a step in the right direction for us, since we'd probably do that anyways.

behlendorf commented:
@edillmann make sure you're running `zdb` on an exported pool, otherwise you may generate false positives.

DeHackEd commented:
@edillmann I'm looking at abd (plain) and don't see this issue on a 2-disk (RAID-0) pool. Examining abd_next shows it's a pretty close to trivial change to what I'm running. space maps are fine and no leaks.

Are you doing any specific testing?

tuxoko commented:
@edillmann
You shouldn't be using abd_next branch or any other abd_*  other than abd. Those are just acting as a bridge for my hosts, and they could be in very buggy or messy states when I'm working on it. But in this moment abd_next is essentially the same as abd. So it's really strange that you could produce the result so easily while I've never seen it myself.
Anyway, could you repull abd branch (not abd_next) and test again? It might be that your branch is stale and in an incomplete state. Thanks.

ahrens commented on include/sys/abd.h:
```diff
@@ -0,0 +1,230 @@
+/*
+ * Convert an linear ABD to normal buffer
+ */
+#define	ABD_TO_BUF(abd)					\
```
This is not C :-)

An "inline static" would be appropriate here, or just drop the assertions.

ahrens commented on include/sys/abd.h:
```diff
@@ -0,0 +1,230 @@
+ * ABD - ARC buffer data
+ * ABD is an abstract data structure for ARC. There are two types of ABD:
+ * linear for metadata and scatter for data.
+ * Their type is determined by the lowest bit of abd_t pointer.
```
Is this still true?

tuxoko replied:
I forgot to modify the comment. Will change that

ahrens commented on include/sys/abd.h:
```diff
@@ -0,0 +1,230 @@
+ * Borrow a linear buffer for an ABD
+ * Will allocate if ABD is scatter
+ */
+#define	abd_borrow_buf(a, n)			\
```
why does this (and friends) need a length parameter ("n")?  Can't you use a->abd_size?

tuxoko replied:
You are right it doesn't need it. I think it was a legacy from older revision. I'll change that.

tuxoko replied:
I'll retract that. They do need a length, becuase the user of these function might be only accessing a part of the buffer.

ahrens replied:
Is that common enough that the performance benefit warrants the added
code?  If so, perhaps adding a 2nd function (abd_borrow_buf_len, or even
the fully-fledged abd_borrow_buf_offset_len) that takes the additional
parameters would make sense.

On Fri, Feb 27, 2015 at 2:10 PM, tuxoko <notifications@github.com> wrote:

> In include/sys/abd.h
> <https://github.com/zfsonlinux/zfs/pull/2129#discussion_r25543144>:
>
> > +{									\
> > +	unsigned int ___ret;						\
> > +	if (ABD_IS_LINEAR(abd))						\
> > +		___ret = bio_map(bio, ABD_TO_BUF(abd) + (off), size);	\
> > +	else								\
> > +		___ret = abd_scatter_bio_map_off(bio, abd, size, off);	\
> > +	___ret;								\
> > +}									\
> > +)
> > +#endif	/* _KERNEL */
> > +
> > +/*
> > + * Borrow a linear buffer for an ABD
> > + * Will allocate if ABD is scatter
> > + */
> > +#define	abd_borrow_buf(a, n)			\
>
> I'll retract that. They do need a length, becuase the user of these
> function might be only accessing a part of the buffer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zfsonlinux/zfs/pull/2129/files#r25543144>.
>

tuxoko replied:
Is not about the performance. Is that we don't want to copy over stuff that we shouldn't.
But I guess we can add a length argument to abd_get_offset and hopefully get around this problem.

ahrens commented:
@tuxoko @behlendorf FYI - I am attempting to port this to illumos.  If you have additional changes, it would be great if it's possible to post them as commits on top of this pull request (rather than redoing the commits you have here), which will make it easier for me to track your new changes.

ahrens commented on module/zfs/zio.c:
```diff
@@ -3028,21 +3112,27 @@ zio_done(zio_t *zio)
+				    zio->io_size);
 			}
 
+			abuf = abd_borrow_buf_copy(adata, asize);
```
adata may be NULL (e.g. if it's a zio_free).  in this case you need to set abuf to NULL; abd_borrow_buf_copy(NULL) won't work.

kernelOfTruth commented on include/sys/abd.h:
```diff
@@ -0,0 +1,230 @@
+
+typedef struct arc_buf_data {
+#ifdef DEBUG_ABD
+	char		pad[PAGE_SIZE];	/* debug, coredumps when accessed */
```
Debug build fails with:

```
error: variably modified 'pad' at file scope
```

tuxoko replied:
Yeah, I already knew this. For now, I think you can just change pad[PAGE_SIZE] to pad[4096].
But I think, I'll just rip out this ugly thing when I push the next revision.

kernelOfTruth replied:
@tuxoko thanks ! that makes the build-bots happy,

I'm just wanting to make sure that potential issues that might crop up later or on some archs are tackled early-on - the build-bots appear to be a worthwhile help in this

DeHackEd commented:
I run my own patch (dehacked/zfs@55ad1f51faecaf4db478409ed3477d094c59cae4, <a href="https://github.com/DeHackEd/zfs/commits/55ad1f51faecaf4db478409ed3477d094c59cae4">commit list</a>) built on ABD plus a bunch of addons that I made or like from the pull request pool. I've had a few hangups lately that I can't tell if they're from Master or ABD.

<pre>INFO: task kswapd0:76 blocked for more than 120 seconds.
     Tainted: P           O 3.14.33 #17
"echo 0 &gt; /proc/sys/kernel/hung_task_timeout_secs" disables this message.
kswapd0         D ffffffffc0d1c1fc     0    76      2 0x00000000
ffff8804177b56e8 0000000000000046 ffff880400000000 ffff8804185ce240
ffff880100000000 ffffea000541e600 ffff8804177b4010 00000000000107c0
ffff8804181ada10 00000000000107c0 00000000000107c0 ffff8804181ada10
Call Trace:
[&lt;ffffffff973df4c5&gt;] schedule+0x6e/0x70
[&lt;ffffffff973df6f7&gt;] schedule_preempt_disabled+0x9/0xb
[&lt;ffffffff973e014b&gt;] __mutex_lock_slowpath+0x1bd/0x213
[&lt;ffffffff970e2718&gt;] ? unfreeze_partials+0x77/0xf9
[&lt;ffffffff973e01bf&gt;] mutex_lock+0x1e/0x2f
[&lt;ffffffffc0c5ac3b&gt;] remove_reference+0x47/0xb6 [zfs]
[&lt;ffffffffc0c5f386&gt;] arc_buf_remove_ref+0xb5/0x111 [zfs]
[&lt;ffffffffc0c64619&gt;] dbuf_rele_and_unlock+0x145/0x229 [zfs]
[&lt;ffffffff9705af0f&gt;] ? update_entity_load_avg+0x16a/0x273
[&lt;ffffffff9705e314&gt;] ? check_preempt_wakeup+0x17f/0x202
[&lt;ffffffff973e01b2&gt;] ? mutex_lock+0x11/0x2f
[&lt;ffffffffc0c64a62&gt;] dbuf_rele+0x25/0x2c [zfs]
[&lt;ffffffffc0c7a763&gt;] dnode_rele+0x6c/0x78 [zfs]
[&lt;ffffffffc0c645a1&gt;] dbuf_rele_and_unlock+0xcd/0x229 [zfs]
[&lt;ffffffffc0bca13e&gt;] ? spl_kmem_cache_free+0x4e/0x18a [spl]
[&lt;ffffffff973e01b2&gt;] ? mutex_lock+0x11/0x2f
[&lt;ffffffff973e01b2&gt;] ? mutex_lock+0x11/0x2f
[&lt;ffffffffc0c64a62&gt;] dbuf_rele+0x25/0x2c [zfs]
[&lt;ffffffffc0c64a72&gt;] dmu_buf_rele+0x9/0xb [zfs]
[&lt;ffffffffc0c981d9&gt;] sa_handle_destroy+0x63/0x9c [zfs]
[&lt;ffffffffc0cdfb9d&gt;] zfs_znode_dmu_fini+0x18/0x27 [zfs]
[&lt;ffffffffc0ce024e&gt;] zfs_zinactive+0xd6/0x10e [zfs]
[&lt;ffffffffc0cd8d87&gt;] zfs_inactive+0x135/0x16a [zfs]
[&lt;ffffffff97062e13&gt;] ? wake_atomic_t_function+0x28/0x28
[&lt;ffffffffc0cf0139&gt;] zpl_evict_inode+0x23/0x27 [zfs]
[&lt;ffffffff97107b3a&gt;] evict+0xaa/0x163
[&lt;ffffffff971081cd&gt;] dispose_list+0x3d/0x4a
[&lt;ffffffff9710821b&gt;] prune_icache_sb+0x41/0x4c
[&lt;ffffffff970f6402&gt;] super_cache_scan+0xfc/0x166
[&lt;ffffffff970b0c73&gt;] shrink_slab_node+0x107/0x166
[&lt;ffffffff970b13eb&gt;] shrink_slab+0x7a/0x11a
[&lt;ffffffff970b429a&gt;] balance_pgdat+0x474/0x607
[&lt;ffffffff970b46c8&gt;] kswapd+0x29b/0x2da
[&lt;ffffffff97062db7&gt;] ? bit_waitqueue+0xb0/0xb0
[&lt;ffffffff9704d94a&gt;] kthread+0xc9/0xd1
[&lt;ffffffff9704d881&gt;] ? kthread_freezable_should_stop+0x60/0x60
[&lt;ffffffff9704d881&gt;] ? kthread_freezable_should_stop+0x60/0x60
     Tainted: P           O 3.14.33 #17
arc_adapt       D ffffffffc0d1bf64     0  3025      2 0x00000000
ffff88008e1eb800 ffff8804185ec010 ffff88040fad4010 00000000000107c0
Call Trace:
[&lt;ffffffff973df6f7&gt;] schedule_preempt_disabled+0x9/0xb
[&lt;ffffffff973dea21&gt;] ? schedule_timeout+0xb9/0xd5
[&lt;ffffffffc0bcf20d&gt;] __cv_timedwait_common+0xb8/0x10d [spl]
[&lt;ffffffffc0bcf270&gt;] __cv_timedwait_interruptible+0xe/0x10 [spl]
[&lt;ffffffffc0cf01bd&gt;] ? zpl_inode_alloc+0x6a/0x6a [zfs]
[&lt;ffffffffc0bcb952&gt;] ? __thread_create+0x122/0x122 [spl]
[&lt;ffffffffc0bcb952&gt;] ? __thread_create+0x122/0x122 [spl]
[&lt;ffffffff9704d881&gt;] ? kthread_freezable_should_stop+0x60/0x60
[&lt;ffffffff9704d881&gt;] ? kthread_freezable_should_stop+0x60/0x60
     Tainted: P           O 3.14.33 #17
l2arc_feed      D ffffffffc0d1c1fc     0  3034      2 0x00000000
ffff88040fafb408 ffffffff9708ea1a ffff88040fafa010 00000000000107c0
Call Trace:
[&lt;ffffffff973df4c5&gt;] schedule+0x6e/0x70
[&lt;ffffffff973e014b&gt;] __mutex_lock_slowpath+0x1bd/0x213
[&lt;ffffffff97062db7&gt;] ? bit_waitqueue+0xb0/0xb0
[&lt;ffffffffc0c6062d&gt;] arc_adjust+0x8d/0x105 [zfs]
[&lt;ffffffffc0c60af5&gt;] arc_kmem_reap_now+0x17/0x98 [zfs]
[&lt;ffffffffc0c60d53&gt;] arc_shrinker_func_scan_objects+0xb/0x1c [zfs]
[&lt;ffffffff970b13eb&gt;] shrink_slab+0x7a/0x11a
[&lt;ffffffff970b398d&gt;] try_to_free_pages+0xa1/0xa8
[&lt;ffffffff970db98a&gt;] alloc_pages_current+0xe7/0x108
[&lt;ffffffff970e1de3&gt;] new_slab+0x5e/0x217
[&lt;ffffffff970e3082&gt;] __slab_alloc+0x2b8/0x3ef
</pre>

tuxoko commented:
@DeHackEd 
There are a lot of deadlocks emerge after the the kmem-rework.
Your log looks like one of them.
There are some partial fix in master or pull request. You can try to cherry pick them.

kernelOfTruth commented on module/zfs/abd.c:
```diff
@@ -0,0 +1,1113 @@
+		offset = sabd->abd_offset + off;
+		abd->abd_offset = offset & (PAGE_SIZE - 1);
+		/* make sure the new abd start as sgl[0] */
+		abd->abd_sgl = &sabd->abd_sgl[offset >> PAGE_SHIFT];
```
@tuxoko 
Since the buildbots are upset about every itty-bitty thing and your patchset, due to the "pad" change are tested more on #3189 , #3190 I'd wanted to share the results with you:

http://buildbot.zfsonlinux.org/builders/ubuntu-12.10-amd64-builder/builds/2883/steps/shell_4
http://buildbot.zfsonlinux.org/builders/ubuntu-12.10-amd64-builder/builds/2883/steps/shell_4/logs/stdio

```
../../module/zfs/abd.c: In function 'abd_get_offset':
../../module/zfs/abd.c:975:43: error: 'PAGE_SHIFT' undeclared (first use in this function)
../../module/zfs/abd.c:975:43: note: each undeclared identifier is reported only once for each function it appears in
../../module/zfs/abd.c: In function 'abd_alloc_scatter':
../../module/zfs/abd.c:1027:38: error: 'PAGE_SHIFT' undeclared (first use in this function)
make[5]: *** [abd.lo] Error 1
```

not sure how to tackle this - since:

PAGE_SHIFT is defined at
include/asm-generic/page.h
and subsequently for each arch on an option-basis differently, but default seems to be 12
arch/*/include/asm/page.h

kernelOfTruth commented on module/zfs/abd.c:
```diff
@@ -0,0 +1,1113 @@
+	abd_t *abd;
+	struct page *page;
+	int i, n = DIV_ROUND_UP(size, PAGE_SIZE);
+	size_t last_size = size - ((n-1) << PAGE_SHIFT);
```
same like line 972, buildbot complains about PAGE_SHIFT being undeclared

kernelOfTruth commented:
@tuxoko this pull-request and the changes from #3115 appear to produce some great results:

https://github.com/zfsonlinux/zfs/pull/3190#issuecomment-84021767

samuelxhu commented:
@tuxoko 
I am running a couple of ZoL-based production servers using ZFS 0.6.3 stable version (the latest one is 0.6.3-1.3 commited by Brian three days ago). My understanding is that your great ABD work will soon be merged right after the release of 0.6.4. Given a lot of tremendous changes such as kmem-rework, large recordsize etc in 0.6.4, i would choose to wait a while until uncertainty settles down. However I am really eager to try/run ABD as soon as possble. 

Therefore may i kindly ask for a back-ported vesion of your ABD work, e.g. ZFS 0.6.3-1.3 stable release. I believe many other ZoL practitioners may have the same wish.

sempervictus commented:
@samuelxhu: i believe that you can fork @tuxoko's github repo, or even this PR itself as a branch and walk the commits back to the state you would like. Far as the major changes in other parts of ZFS go - ARC is generally a "safer" place to make large changes IMO, since it's not responsible for committing data to disk. We have several workloads running production ZVOLs with the kmem rework code in the base of our patch stack, and it works just fine, but we still aren't doing datasets this way for heavily loaded systems since direct reclaim issues keep popping up like methylated gophers (at least in the issues tracker, we've been lucky enough to avoid most of those).

samuelxhu commented:
@sempervictus: Thanks for the suggestions. Definitely I will try as what you described. My concern is, this may not be a clean way for production, given that ABD itself is sophisticated. 
@tuxoko: As you may already noted, Brian is actively maintain 0.6.3 stable release, and it is now widely used in production. I really hope you could find time (and interest of course) to issue a ABD patch for it. Myself would appreciate it very much.

kernelOfTruth commented:
just a heads-up:

latest upstream/master branch underwent some significant changes due to commit https://github.com/zfsonlinux/zfs/commit/d683ddbb7272a179da3918cc4f922d92a2195ba2

Illumos 5314 - Remove "dbuf phys" db->db_data pointer aliases in ZFS

which need to be addressed when again updating this pull-request to work with latest code

tuxoko commented:
@kernelOfTruth 
Thanks for the notice. It seems to require quite some work.
Btw, I'm currently working on a major enhancement for the abd, but it seems it will have to wait for a bit.

sempervictus commented:
I just hit this as well - combing through the merge in that commit, looks like this might affect a number of outstanding PRs.

On a separate note, we've been successfully running test and now production workloads (SCST iSCSI from ZVOLs and NFS/SMB/etc from datasets) using #3216 (and preceding PRs). The whole stack shows very encouraging results - significant reduction of IOWait on heavily contended systems, much less CPU time allocated to ARC related kernel tasks across the board, and a rather noticeable response time decrease at the higher level XenServer VMs atop the exports (which in that case run on 6.2.5 and limited to 512b iSCSI sector sizing resulting in abusive IO access patterns). 

Thank you @tuxoko for maintaining this as an active WIP while master shifts around you.

tuxoko commented:
Hi guys,
I've update the pull request. The last version is still accessible with branch abd_archive00.

This update includes a major enhancement: scatter buffer for metadata.
The old version only applied scatter buffer to user data. So for metadata heavy use case, the memory fragmentation is still quite an issue. This new version applies scatter buffer to indirect blocks(bp level > 0) and dnode arrays(DMU_OT_DNODE). I'm planning to add support to ZAP as well. After that, I believe it can cover most metadata heavy workload.

Note that this version may not be as stable as the previous ones, so please test it hard.

kernelOfTruth commented:
@tuxoko Thanks a lot for the quick & extensive update !

I'm wondering what changes this will make for performance and under high load

behlendorf commented:
@tuxoko thanks for refreshing this (again!).  I'd like to plan to get this in the next tag.  Right after the large block patches in #3363 are merged.  I'm hoping to merge the large block patch in the next week or two so it would be great if there was a version of this patch rebased on the #3363 branch ready to merge.  These would be the two of the major changes for the next tag.

kpande commented:
if you rebase this with large_blocks I can begin my own testing of these changes.

tuxoko commented:
@behlendorf Great! I'll see what I need to change to accommodate the large block patches.

behlendorf commented:
@tuxoko All green from the buildbot, that's very encouraging!

behlendorf commented:
@tuxoko the large block patches have been merged to master.  If you could refresh this patch stack to apply cleanly, and handle the larger block size, we can focus on getting this in next.  I'd really like to push to get this in the next tag.

tuxoko commented:
@behlendorf 
I'm working on it. Also, I'm doing a full review which I haven't done since the first version of the ABD. So it would take some time. But I should be able to finish it in a week or two.

don-brady commented:
@tuxoko
I've just started to review this (huge thanks for taking on the work). Hope to test as well soon. Some initial comments:

1. In _abd_alloc_scatter(), when checking for the minimum threshold, can we check instead for less than PAGE_SIZE*2 to also avoid the 5120, 6144 and 7168 bufsize cases?

2. the abd_borrow_buf() macro makes it difficult to uniquely trace that function since it turns into a generic zio_buf_free() call. Could this be made into an inline function instead (even inline functions can be traced). Same for abd_return_buf().

tuxoko commented:
@don-brady 

1. There's nothing preventing us from doing that. I chose PAGE_SIZE be cause it would limit the wasted memory to less than 100%, like kmalloc. But if no one objects, I will change it to 2*PAGE_SIZE.

2. I was planing to change it anyway, because it uses non-standard statement expression.

behlendorf commented:
@tuxoko @don-brady 

1. We need to be careful to avoid making any assumptions about PAGE_SIZE.  We can't count on it being 4k since we want this code to work on aarch and ppc architectures where alternate page sizes are entirely valid.  I'd suggest that if there's some doubt about what the right cutoff here should be for an optimization we make it module option for now until the best value can be settled on.

don-brady commented:
@tuxoko @behlendorf

Overall seems to be working fine alongside the large block changes (local manual merge).

side note:  When observing _abd_alloc_scatter() calls while using large record sizes (16M), in addition to the expected large allocations, I see an unexpected number of 12K and 16K scatter allocations coming through.  These appear to be mostly coming from vdev io aggregations of 4K metadata (note my vdev ashift is 12).   Just curious if we want/need these transient metadata aggregation buffers to be scatter? I don’t think it’s a problem, I just wasn’t expecting to see them.

behlendorf commented:
@don-brady I think this is best.  If we're going to allow the aggregation then my gut feeling would be that it makes the most sense to use the scatter lists for this.  Allocating pages and mapping them atomically to be zeroed/copied  (as is done) should be very efficient and avoids any potential lock contention is a shared allocator like the slab.  We can always revisit this is that turns out not to be the case!

As another related aside it's never been 100% clear that the existing vdev queue aggregation under Linux is always a win.  In many cases this _might_ be wasted work which the block device elevator would  have efficiently taken care of.  Although there are certainly exceptions which may make the whole thing worth while like the gap spanning.  But that's neither here nor there for this issue!

tuxoko commented:
@don-brady 
If you're using the newest version. What you see could be the indirect blocks. In my test, I see a lot of them being 16K. And I added support for using scatter buffer for them.

@behlendorf 
Yes, I always wonder why we would need aggregation when we already have linux elevator.

sempervictus commented:
@tuxoko: it appears that this stack is somehow affecting space calculation for L2ARC devices. Could you take a look @ #3400 and weigh in? I'm now seeing the problem off of abd_next without any of the alterations brought by #3115 or additional changes from @kernelOfTruth's work.

tuxoko commented:
Hi guys, I've just updated this pull request.
This pull request add the support for large block by using chained scatterlist, and a bunch of other misc stuff. The commit history might look a bit messy, because I want to retain what I added with respect to old version in case @ahrens want to know about them.

I'll submit a new and rebased pull request after I see some result from the buildbot.

behlendorf commented:
@tuxoko nice work.  Hopefully latter this week I'll have some time to start on a careful review.  In the meanwhile I'll instruct the buildbot to go through and test all the patches in this stack.

DeHackEd commented:
From the current head I ended up need this in order to build it properly:
DeHackEd/zfs@092bab58e7a3ac5f413a6a3cc670320e364b84c4

behlendorf commented:
Closing, refreshed as pull request #3441.  Let's move addition review and comments in to the updated version.
